content/en/blog/_posts/2023-03-08-nodeport-dynamic-and-static-allocation.md

```diff
@@ -0,0 +1,162 @@
+to a specific port of all nodes in the cluster, thus redirecting traffic from the
+outside to the inside of the cluster.
+
+## How NodePort's ports are allocated? 
+
+When a NodePort Service is created, its corresponding port is allocated in two ways.
+
+- **Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`
+  is not specified, the Kubernetes control plane will automatically allocate an unused port 
+  to it at creation time.
+
+- **Static** : In addition to the dynamic auto-assignment described above, you can also
+  explicitly assign a port that is within the nodeport port range configuration.
+
+the `service.spec.ports.nodePort` must be unique for each NodePort Service across the cluster.
+Attempting to create a NodePort type of Service with a port already allocated will return an error.
+
+## Why do you need to reserve ports of NodePort Service? 
+Sometimes, you may want to have a NodePort Service running on well-known ports
+so that other components and users inside o r outside the cluster can use them.
+
+In some complex cluster deployments with a mix of K8S nodes and non-K8S nodes, it may be
+necessary to use some pre-defined ports for communication. In particular, some fundamental
+components cannot rely on the VIPs which provided by LoadBalancer services,
+because the load balancer that provides the VIPs may also rely on these fundamental components.
+
+Now suppose we need to expose a Minio object storage service on K8S to services located on non-K8S nodes,
+and the agreed port is `30009`, we need to create a Service as follows:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    k8s-app: minio
+  name: minio
+  namespace: kube-system
+spec:
+  ports:
+  - name: http
+    nodePort: 30009
+    port: 9000
+    protocol: TCP
+    targetPort: 9000
+  selector:
+    app: minio
+  type: NodePort
+```
+
+However, as mentioned before, if the port `30009` required for the minio Service is not reserved,
+and another NodePort Service is created and dynamically allocated before or concurrently with the minio
+Service, `30009` can be allocated to those NodePort Service, while the minio Service will fail to be
+created due to a port conflict.
+
+## How can you avoid NodePort Service port conflicts? 
+In the Kubernetes 1.24, Service ClusterIP had already divided the CIDRs into two blocks, using different
+allocation policies to reduce the risk of conflicts. In Kubernetes 1.27, a similar policy can be adopted
+for NodePort Services. You can enable a new feature gate `ServiceNodePortStaticSubrange`. Turning this on allows you
+to use a different port allocation strategy for NodePort Services, and reducing the risk of collision.
+
+The port range for `NodePort` will be divided, based on the formula `min(max(16, nodeport-size / 32), 128)`,
+which can be described as _never less than 16 or more than 128 with a graduated step between them_.
```

Comment:
This doesn't make sense. What does the outcome of the formula determine?

If my cluster is running with the configuration `--service-node-port-range=65520-65535` and I enable this feature gate, what happens?

(“nobody would ever do that” doesn't feel like a realistic response, given that there's a similar example just a bit later)

Reply:
I modified some wording to explain the meaning of the formula output.

If `--service-node-port-range=65520-65535` is configured and the size of the node port range is set to 15, which is less than 16, then the size of the static port range should be set to 0. This means that all ports will be allocated using dynamic policy.